packages/liveblocks-react/src/liveblocks.tsx

```diff
@@ -1179,6 +1190,77 @@ const _useUserThreads_experimental: TypedBundle["useUserThreads_experimental"] =
 const _useUserThreadsSuspense_experimental: TypedBundle["suspense"]["useUserThreads_experimental"] =
   useUserThreadsSuspense_experimental;
 
+function useSyncStatus_withClient(
+  client: OpaqueClient,
+  options?: UseSyncStatusOptions
+): SyncStatus {
+  // Normally the Rules of Hooks™ dictate that you should not call hooks
+  // conditionally. In this case, we're good here, because the same code path
+  // will always be taken on every subsequent render here, because we've frozen
+  // the value.
+  /* eslint-disable react-hooks/rules-of-hooks */
+  const smooth = useInitial(options?.smooth ?? false);
+  if (smooth) {
+    return useSyncStatusSmooth_withClient(client);
+  } else {
+    return useSyncStatusImmediate_withClient(client);
+  }
+  /* eslint-enable react-hooks/rules-of-hooks */
+}
+
+function useSyncStatusImmediate_withClient(client: OpaqueClient): SyncStatus {
+  return useSyncExternalStore(
+    client.events.syncStatus.subscribe,
+    client.getSyncStatus,
+    client.getSyncStatus
+  );
+}
+
+function useSyncStatusSmooth_withClient(client: OpaqueClient): SyncStatus {
+  const getter = client.getSyncStatus;
+  const [status, setStatus] = React.useState(getter);
+  const oldStatus = useLatest(getter());
+
+  React.useEffect(() => {
+    let timeoutId: ReturnType<typeof setTimeout>;
+    const unsub = client.events.syncStatus.subscribe(() => {
+      const newStatus = getter();
+      if (
+        oldStatus.current === "synchronizing" &&
+        newStatus === "synchronized"
+      ) {
+        // Delay delivery of the "synchronized" event
+        timeoutId = setTimeout(() => setStatus(newStatus), config.SMOOTH_DELAY);
+      } else {
+        clearTimeout(timeoutId);
+        setStatus(newStatus);
```

Comment:
Do you know if it is possible to also use `useSyncExternalStore` for this hook too? I am not a 100% certain if it is applicable in this particular case I think it could be possible for "[tearing](https://github.com/reactwg/react-18/discussions/69)" to happen if one uses `useEffect` and `useState` to synchronize with an external store.

Reply:
That would be nice indeed, but I don't know what would be the best approach here. The state being observed here is deliberately decoupled from the external store (via the timeout). We could surely read the immediate status value from the external store and then use it to compute a delayed state transition, i.e. in pseudo-code:

```
const immediateStatus = useSyncStatusImmediately();  // will already be uSES-based
const status = useDelayedStatus(immediateStatus);  // will use timer to smoothen
```

…but in this case, we're already too late because using the `useSyncExternalStore`-based approach will mean that this component will re-render a lot (up to 60 times per second).

Another approach could be to somehow try to shoehorn this into `useSyncExternalStore`, by being smart about the subscription event, and trying to hide/defer certain status change events, but I think fundamentally this is impossible because by the time those events fire, reading the status synchronously will no longer match the status we want to display.

The only solution I see that would allow us to use `useSyncExternalStore` here is to make the external store _be_ the deferred/smooth status, which would require moving this logic into the client instead.

Or can you think of another approach here? Open to any ideas!

Reply:
I can see how an implementation might be challenging 🤔 . Let's invest more time on this if there are bug reports from user concerning tearing!